In [29]:
import pandas as pd 
import numpy as np 
import seaborn as sns 
import matplotlib as plt 
import sklearn
from sklearn.model_selection import train_test_split

In [30]:
# Reading all the relevant Datasets

customers = pd.read_csv('olist_customers_dataset.csv')
order_items = pd.read_csv('olist_order_items_dataset.csv')
products = pd.read_csv('olist_products_dataset.csv')
orders = pd.read_csv('olist_orders_dataset.csv')

In [31]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [32]:
order_items['order_item_id'].value_counts()

1     98666
2      9803
3      2287
4       965
5       460
6       256
7        58
8        36
9        28
10       25
11       17
12       13
13        8
14        7
15        5
16        3
17        3
18        3
19        3
20        3
21        1
Name: order_item_id, dtype: int64

In [33]:
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [34]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

# Data Check/update step 1

In [35]:
# Merging datasets to make a centralized one with all the relevant data

dataframe = pd.merge(products, order_items, on = 'product_id')

# Data Check / update step 2

In [36]:
dataframe = dataframe.dropna()
dataframe = dataframe.drop(columns = ['order_id', 'order_item_id', 'seller_id', 'shipping_limit_date'])


# Data Check / update step 3

In [37]:
dataframe = dataframe.drop(columns = ['product_category_name', 'product_id', 'product_photos_qty'])
dataframe

product_name_lenght  product_description_lenght  product_weight_g  \
0                      40.0                       287.0             225.0   
1                      44.0                       276.0            1000.0   
2                      46.0                       250.0             154.0   
3                      27.0                       261.0             371.0   
4                      37.0                       402.0             625.0   
...                     ...                         ...               ...   
112645                 60.0                       156.0             700.0   
112646                 60.0                       156.0             700.0   
112647                 60.0                       156.0             700.0   
112648                 60.0                       156.0             700.0   
112649                 58.0                       309.0            2083.0   

        product_length_cm  product_height_cm  product_width_cm   price  \
0                    16.0               10.0              14.0   10.91   
1                    30.0               18.0              20.0  248.00   
2                    18.0                9.0              15.0   79.80   
3                    26.0                4.0              26.0  112.30   
4                    20.0               17.0              13.0   37.90   
...                   ...                ...               ...     ...   
112645               31.0               13.0              20.0   29.90   
112646               31.0               13.0              20.0   29.90   
112647               31.0               13.0              20.0   29.90   
112648               31.0               13.0              20.0   29.90   
112649               12.0                2.0               7.0  107.50   

        freight_value  
0                7.39  
1               17.99  
2                7.82  
3                9.54  
4                8.29  
...               ...  
112645          16.11  
112646          23.28  
112647          23.28  
112648          23.28  
112649          27.05  

[111046 rows x 8 columns]

In [8]:
train, test = train_test_split(dataframe, test_size = 0.2)
train.dtypes

product_name_lenght           float64
product_description_lenght    float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
price                         float64
freight_value                 float64
dtype: object

In [21]:
# Splitting my dataset into a train and test group

x_train = train.drop(columns = ['price'])
y_train = train['price']

x_test = test.drop(columns = ['price'])
y_test = test['price']

In [10]:
x_train

product_name_lenght  product_description_lenght  product_weight_g  \
14993                  23.0                       591.0             400.0   
32559                  38.0                       859.0            1100.0   
100117                 58.0                      3006.0            1000.0   
92101                  58.0                       678.0             150.0   
94064                  47.0                      1801.0            1550.0   
...                     ...                         ...               ...   
33030                  54.0                       245.0            1383.0   
51053                  60.0                       530.0            1225.0   
91495                  52.0                        85.0             200.0   
4975                   59.0                       207.0             300.0   
68884                  52.0                       747.0           14500.0   

        product_length_cm  product_height_cm  product_width_cm  freight_value  
14993                24.0               16.0              25.0          15.42  
32559                20.0               25.0              25.0          13.21  
100117               53.0                8.0              18.0          16.52  
92101                17.0                8.0              14.0          18.31  
94064                20.0               50.0              20.0          18.14  
...                   ...                ...               ...            ...  
33030                50.0               10.0              40.0          12.13  
51053                43.0               38.0              37.0           7.90  
91495                16.0                2.0              11.0          18.23  
4975                 40.0               25.0              25.0          12.61  
68884                40.0               60.0              40.0          65.50  

[88836 rows x 7 columns]

In [19]:
# Building my Linear Regression Model

from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(X = x_train, y =  y_train)

pred = model.predict(x_test)

rmse = np.mean((np.sqrt((pred - y_test)**2)))

print(rmse)

78.8865731185267


In [13]:
np.mean(y_test)

120.11395317424585

In [17]:
np.mean(dataframe)

product_name_lenght             48.775877
product_description_lenght     787.866335
product_weight_g              2099.979657
product_length_cm               30.211003
product_height_cm               16.626362
product_width_cm                23.041956
price                          120.762337
freight_value                   20.025149
dtype: float64

In [14]:
dataframe.to_csv('dataset_pre_processed1.csv')

# Summary

(a) What 3 checks or updates to clean your data did you implement and why?

A: First I merged two datasets based on a unique column - product ID. I did this as the both datasets had information that was relevant for me and I needed all of that information in one centralized place. Next, I checked if there were null values and dropped all the rows that had null values. I did this to ensure that all the data fiels in my dataset were complete so that my models won't error due to missing values. Third, I dropped all columns that had strings and were also irrelevant for my analysis. 

(b) What are example means / averages in your dataset, and what is your hypothesis on eventual model results?

A: The average price of the products listed on my dataset was 120 and the average freight value was 20. My hypothesis was to predict the price of online products sold in Brazil based on the product specifications. I was expecting the model results to have a rmse of around 25. 

(c) How would you explain your H2O.ai results and your own model results, and was your hypothesis right or wrong?

A: Though my dataset had over 110,000 rows both the H2o model and my own model were not able to accurately predict price. I was adament to find a dataset on online orders & delivery to build a model since our project is tackling last mile delivery problems. However, I believe the dataset I found lacked sufficient features to predict the price accurately. I see this in hindsight since the RMSE of both models are higher than my hypothesis exected. 